### PR TITLE
Bug: Handling of symbols when used as deps incorrectly to create error message results in an unrelated TypeError: Cannot convert a Symbol value to a string 

### LINKED ISSUE #19848: Bug: Handling of symbols when used as deps incorrectly to create error message results in an unrelated TypeError: Cannot convert a Symbol value to a string
React version: 16.3.1

## Steps To Reproduce

1. Click the button in this Code Sandbox https://codesandbox.io/s/blissful-sun-e0lle?file=/src/App.js

## The current behavior

The wrong error is generated

## The expected behavior

An error should still be generated but the error message should be correct. The problem is that if you do `[Symbol('...')].join(',')` JavaScript will freak out which is what happens if you put symbols incorrectly in the deps to hooks.

### PR DESCRIPTION
**Pull request by omarsy**
Fixes #19848
## Summary
image

### CODE REVIEW

File: packages/react-reconciler/src/ReactFiberHooks.new.js [hunk 1]
```diff
@@ -194,6 +194,20 @@ let hookTypesUpdateIndexDev: number = -1;
 // When true, such Hooks will always be "remounted". Only used during hot reload.
 let ignorePreviousDependencies: boolean = false;
 
+function joinArray(collection: Array<mixed>, separator: string = ''): string {
```
Comment (omarsy) [inline packages/react-reconciler/src/ReactFiberHooks.new.js:197]: Maybe we can move this function in another file ?

**Comment by leidegre** [comment]
Since you did make seperate function you might as well apply it recursively. Otherwise, the error will still happen if you for some reason have `''+[[Symbol()]]` (nested arrays with symbols) and since this is a helper function for a very specific code path I would nest the function under the `if (__DEV__)` statement unless there's something in the coding guidelines against this because it more clearly communicates that this function exists to solve a very specific problem.

With that in mind, I think a comment about why it's done like this is prudent.

**Comment by omarsy** [comment]
bump

**Comment by eps1lon** [comment]
Note that this is also helps with nullish deps. `[null, undefined].join(', ') === ", "` while the proposed implementation results in `"null, undefined"`. Though we might want to special case strings so that `null` and `"null"` are displayed differently:

```ts
/**
 * Supports Symbol dependencies
 * Better display for nullish dependencies
 */
function stringifyHookDependencies(dependencies) {
  return dependencies
    .map((dependency) => {
      if (typeof dependency === "string") {
        return `"${dependency}"`;
      }

      return String(dependency);
    })
    .join(", ");
}
```

**Pull request closed without merging**